include/control.hpp

```diff
@@ -0,0 +1,260 @@
+#pragma once
+#include <algorithm>
+#include <chrono>
+#include <cstdint>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <vector>
+#include <mutex>
+
+#ifdef DEBUG
+#include <ctime>
+#include <sstream>
+#endif
+
+#define RPM_TO_RADS_SEC 0.10472
+
+namespace Control {
+
+/* classes */
+class PidController;
+class SkidRobotMotionController;
+class AlphaBetaFilter;
+
+/* datatypes */
+typedef enum {
+  OPEN_LOOP = 0,
+  TRACTION_CONTROL = 1,
+  INDEPENDENT_WHEEL = 2,
+} robot_motion_mode_t;
+
+const uint8_t NUM_MOTION_MODES = 3;
+
+struct angular_scaling_params {
+  float a_coef;
+  float b_coef;
+  float c_coef;
+  float min_scale_val;
+  float max_scale_val;
+};
+
+struct robot_velocities {
+  float linear_velocity;
+  float angular_velocity;
+};
+
+struct motor_data {
+  float fl;
+  float fr;
+  float rl;
+  float rr;
+};
+
+struct robot_geometry {
+  float intra_axle_distance;
+  float wheel_base;
+  float wheel_radius;
+  float center_of_mass_x_offset;
+  float center_of_mass_y_offset;
+};
+
+struct pid_gains {
+  double kp;
+  double ki;
+  double kd;
+};
+
+struct pid_outputs {
+  std::string name;
+  double time;
+  float dt;
+  float pid_output;
+  float error;
+  float integral_error;
+  float delta_error;
+  float target_value;
+  float measured_value;
+  double kp;
+  double ki;
+  double kd;
+};
+
+struct pid_output_limits {
+  float posmax;
+  float negmax;
+};
+
+/* useful functions */
+motor_data computeSkidSteerWheelSpeeds(robot_velocities target_velocities,
+                                       robot_geometry robot_geometry);
+
+robot_velocities limitAcceleration(robot_velocities target_velocities,
+                                   robot_velocities measured_velocities,
+                                   robot_velocities delta_v_limits, float dt);
+
+robot_velocities scaleAngularCommand(robot_velocities target_velocities,
+                                     robot_velocities measured_velocities,
+                                     angular_scaling_params scaling_params);
+
+robot_velocities computeVelocitiesFromWheelspeeds(
+    motor_data wheel_speeds, robot_geometry robot_geometry);
+
+}  // namespace Control
+
+class Control::PidController {
+ public:
+  /* constructors */
+  PidController(pid_gains pid_gains, std::string name);
+  PidController(pid_gains pid_gains, pid_output_limits pid_output_limits,
+                std::string name);
+
+  void setGains(pid_gains pid_gains);
+  pid_gains getGains();
+
+  void setOutputLimits(pid_output_limits pid_output_limits);
+  pid_output_limits getOutputLimits();
+
+  void setIntegralErrorLimit(float error_limit);
+  float getIntegralErrorLimit();
+
+  pid_outputs runControl(float target, float measured);
+
+  void writePidDataToCsv(std::ofstream &log_file, pid_outputs data);
```

Comment:
/* brief
*
*/ are missing for these function. Intellisense would not be able to disaply any useful description for these